uuid.go

```diff
@@ -74,44 +78,58 @@ func Parse(s string) (UUID, error) {
 		}
 		s = s[9:]
 
+		// Ensure proper dash placement
+		if s[8] != '-' || s[13] != '-' || s[18] != '-' || s[23] != '-' {
+			return uuid, errors.New("invalid UUID format")
+		}
+		s = s[1:]
+
 	// {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
 	case 36 + 2:
+		// Ensure proper dash placement
+		if s[8] != '-' || s[13] != '-' || s[18] != '-' || s[23] != '-' {
+			return uuid, errors.New("invalid UUID format")
+		}
 		s = s[1:]
 
 	// xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 	case 32:
-		var ok bool
+		// Ensure proper length and valid characters
 		for i := range uuid {
-			uuid[i], ok = xtob(s[i*2], s[i*2+1])
-			if !ok {
+			uuid[i] = s[i]
+			if !isValidHexChar(s[i]) {
 				return uuid, errors.New("invalid UUID format")
 			}
 		}
+
 		return uuid, nil
+
 	default:
 		return uuid, invalidLengthError{len(s)}
 	}
-	// s is now at least 36 bytes long
-	// it must be of the form  xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
-	if s[8] != '-' || s[13] != '-' || s[18] != '-' || s[23] != '-' {
-		return uuid, errors.New("invalid UUID format")
-	}
+
+	// Validate the UUID parts
 	for i, x := range [16]int{
 		0, 2, 4, 6,
 		9, 11,
 		14, 16,
 		19, 21,
 		24, 26, 28, 30, 32, 34,
 	} {
-		v, ok := xtob(s[x], s[x+1])
-		if !ok {
+		if !isValidHexChar(s[x]) || !isValidHexChar(s[x+1]) {
```

Comment:
This block is both slower and produces the wrong results.  both s[x] and s[x+1] are required to determine the value of uuid byte `i`.  The `xtob` function, which handles this properly is very optimized.